src/System.Private.CoreLib/src/System/Diagnostics/Eventing/EventPipeEventDispatcher.cs

```diff
@@ -0,0 +1,162 @@
+        {
+            if (command == EventCommand.Update && enable)
+            {
+                lock (m_dispatchControlLock)
+                {
+                    // Add the new subscription.  This will overwrite an existing subscription for the listener if one exists.
+                    m_subscriptions[eventListener] = new EventListenerSubscription(matchAnyKeywords, level);
+
+                    // Commit the configuration change.
+                    CommitDispatchConfiguration();
+                }
+            }
+            else if (command == EventCommand.Update && !enable)
+            {
+                RemoveEventListener(eventListener);
+            }
+        }
+
+        internal void RemoveEventListener(EventListener listener)
+        {
+            lock (m_dispatchControlLock)
+            {
+                // Remove the event listener from the list of subscribers.
+                if (m_subscriptions.ContainsKey(listener))
+                {
+                    m_subscriptions.Remove(listener);
+                }
+
+                // Commit the configuration change.
+                CommitDispatchConfiguration();
+            }
+        }
+
+        private void CommitDispatchConfiguration()
+        {
+            // Ensure that the dispatch task is stopped.
+            // This is a no-op if the task is already stopped.
+            StopDispatchTask();
+
+            // Stop tracing.
+            // This is a no-op if it's already disabled.
+            EventPipeInternal.Disable();
+
+            // Check to see if tracing should be enabled.
+            if (m_subscriptions.Count <= 0)
+            {
+                return;
+            }
+
+            // Start collecting events.
+            EventKeywords aggregatedKeywords = EventKeywords.None;
+            EventLevel highestLevel = EventLevel.LogAlways;
+
+            foreach (EventListenerSubscription subscription in m_subscriptions.Values)
+            {
+                aggregatedKeywords |= subscription.MatchAnyKeywords;
+                highestLevel = (subscription.Level > highestLevel) ? subscription.Level : highestLevel;
+            }
+
+            EventPipeProviderConfiguration[] providerConfiguration = new EventPipeProviderConfiguration[]
+            {
+                new EventPipeProviderConfiguration(RuntimeEventSource.EventSourceName, (ulong) aggregatedKeywords, (uint) highestLevel)
+            };
+
+            EventPipeInternal.Enable(null, 1024, 1, providerConfiguration, 1);
+
+            // Start the dispatch task.
+            StartDispatchTask();
+        }
+
+        private void StartDispatchTask()
+        {
+            Debug.Assert(Monitor.IsEntered(m_dispatchControlLock));
+
+            if (m_dispatchTask == null)
+            {
+                m_stopDispatchTask = false;
+                m_dispatchTask = Task.Run(DispatchEventsToEventListeners);
```

Comment:
If this is long running then you don't want to use Task.Run as that'll tie up a thread pool thread. Better to just spawn a real thread.

Reply:
Yeah, it's likely to be long running.  Are you concerned about thread pool thread spin-up?  I don't imagine that this will hurt, but at the same time, creating a new thread won't hurt either - I'd expect them to be about the same.  That said, I can look at using a new thread instead.

Reply:
That’s right best practice here is to either use TaskCreationOptions.LongRunning (this spins up a thread under the covers) or spin up a thread yourself.

Reply:
Fair enough.  I've updated the task creation code to mark it as long running, so it will just create a new thread under the covers.